File: backend/app/schemas/dag.py
"""Schemas for DAG system."""

from typing import Dict, List, Optional
from uuid import UUID

from pydantic import BaseModel, Field


class DagNodePosition(BaseModel):
    """Schema for node position in the UI."""

    x: float = Field(..., description="X coordinate")
    y: float = Field(..., description="Y coordinate")


class DagNodeBase(BaseModel):
    """Base schema for DAG node."""

    type: str  # source, destination, transformer, entity
    name: str
    config: Optional[Dict] = None
    position: Optional[DagNodePosition] = None

    # One of these will be set based on type
    source_id: Optional[UUID] = None
    destination_id: Optional[UUID] = None
    transformer_id: Optional[UUID] = None
    entity_definition_id: Optional[UUID] = None


class DagNodeCreate(DagNodeBase):
    """Schema for creating a DAG node."""

    pass


class DagNodeUpdate(DagNodeBase):
    """Schema for updating a DAG node."""

    pass


class DagNode(DagNodeBase):
    """Schema for a DAG node."""

    id: UUID
    dag_id: UUID
    organization_id: UUID
    created_by_email: str
    modified_by_email: str

    class Config:
        """Pydantic config."""

        from_attributes = True


class DagEdgeBase(BaseModel):
    """Base schema for DAG edge."""

    from_node_id: UUID
    to_node_id: UUID


class DagEdgeCreate(DagEdgeBase):
    """Schema for creating a DAG edge."""

    pass


class DagEdgeUpdate(DagEdgeBase):
    """Schema for updating a DAG edge."""

    pass


class DagEdge(DagEdgeBase):
    """Schema for a DAG edge."""

    id: UUID
    dag_id: UUID
    organization_id: UUID
    created_by_email: str
    modified_by_email: str

    class Config:
        """Pydantic config."""

        from_attributes = True


class SyncDagDefinitionBase(BaseModel):
    """Base schema for sync DAG definition."""

    name: str
    description: Optional[str] = None
    sync_id: UUID


class SyncDagDefinitionCreate(SyncDagDefinitionBase):
    """Schema for creating a sync DAG definition."""

    nodes: List[DagNodeCreate]
    edges: List[DagEdgeCreate]


class SyncDagDefinitionUpdate(SyncDagDefinitionBase):
    """Schema for updating a sync DAG definition."""

    nodes: Optional[List[DagNodeCreate]] = None
    edges: Optional[List[DagEdgeCreate]] = None


class SyncDagDefinition(SyncDagDefinitionBase):
    """Schema for a sync DAG definition."""

    id: UUID
    organization_id: UUID
    created_by_email: str
    modified_by_email: str
    nodes: List[DagNode]
    edges: List[DagEdge]

    class Config:
        """Pydantic config."""

        from_attributes = True

File: frontend/src/types/dag.ts
import { Node, Edge } from 'reactflow';

export interface DagNode {
  id: string;
  type: 'SOURCE' | 'DESTINATION' | 'ENTITY' | 'TRANSFORMER';
  name: string;
  shortName?: string;
  config?: Record<string, any>;
  sourceId?: string;
  destinationId?: string;
  transformerId?: string;
  entityDefinitionId?: string;
}

export interface DagEdge {
  id: string;
  fromNodeId: string;
  toNodeId: string;
}

export interface DagDefinition {
  id: string;
  name: string;
  description?: string;
  syncId: string;
  nodes: DagNode[];
  edges: DagEdge[];
}

export interface FlowNode extends Node {
  data: {
    name: string;
    shortName?: string;
    config?: Record<string, any>;
    sourceId?: string;
    destinationId?: string;
    transformerId?: string;
    entityDefinitionId?: string;
  };
}

export interface FlowEdge extends Edge {
  type: 'button';
}

// Conversion utilities
export const toFlowNodes = (nodes: DagNode[]): FlowNode[] => 
  nodes.map(node => ({
    id: node.id,
    type: node.type.toLowerCase(),
    data: {
      name: node.name,
      shortName: node.shortName,
      config: node.config,
      sourceId: node.sourceId,
      destinationId: node.destinationId,
      transformerId: node.transformerId,
      entityDefinitionId: node.entityDefinitionId,
    },
    position: { x: 0, y: 0 }, // Will be set by layout
  }));

export const toFlowEdges = (edges: DagEdge[]): FlowEdge[] =>
  edges.map(edge => ({
    id: edge.id,
    source: edge.fromNodeId,
    target: edge.toNodeId,
    type: 'button',
  }));

export const toDagNodes = (nodes: FlowNode[]): DagNode[] =>
  nodes.map(node => ({
    id: node.id,
    type: node.type.toUpperCase() as DagNode['type'],
    name: node.data.name,
    shortName: node.data.shortName,
    config: node.data.config,
    sourceId: node.data.sourceId,
    destinationId: node.data.destinationId,
    transformerId: node.data.transformerId,
    entityDefinitionId: node.data.entityDefinitionId,
  }));

export const toDagEdges = (edges: FlowEdge[]): DagEdge[] =>
  edges.map(edge => ({
    id: edge.id,
    fromNodeId: edge.source,
    toNodeId: edge.target,
  })); 